### ISSUE TITLE
Creating alternator table fails with Scylla 5.4.0 and consistent-topology-changes

### ISSUE THREAD

**Issue by Michal-Leszczynski** [scylladb/scylladb#16332]
With Scylla 5.2 creating table with [example script](https://opensource.docs.scylladb.com/stable/alternator/getting-started.html#running-some-simple-scripts) from docs ends successfully, but with Scylla 5.4.0 it fails with the following error:
```
[miles@fedora ~]$ python alt-table.py 
Traceback (most recent call last):
  File "/home/miles/alt-table.py", line 5, in <module>
    dynamodb.create_table(
  File "/usr/local/lib/python3.11/site-packages/boto3/resources/factory.py", line 580, in do_action
    response = action(self, *args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/boto3/resources/action.py", line 88, in __call__
    response = getattr(parent.meta.client, operation_name)(*args, **params)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/botocore/client.py", line 535, in _api_call
    return self._make_api_call(operation_name, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/botocore/client.py", line 980, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.errorfactory.InternalServerError: An error occurred (InternalServerError) when calling the CreateTable operation (reached max retries: 9): Internal server error: data_dictionary::no_such_keyspace (Can't find a keyspace alternator_usertable)
```

Is this a config issue?
- used scylla.yaml: [scylla.zip](https://github.com/scylladb/scylladb/files/13600547/scylla.zip)
- cluster topology: 2 dcs, 3 nodes each

@nuivall

**Comment by nuivall** [comment]
It works for me on master (5.5.0~dev-0.20231207.92d61def5748), looks like either some general misconfiguration or nasty regression. 

Can you try to repeat this on fresh cluster or locally with exact the same version? Potentially mutli-dc could be a trigger as well.

One thing alternator is doing differently than CQL is that it creates keyspace and table at the same time, if this is not a regression but rather "it's always been there" issue then potentially consistent schema management feature could help, not sure since which version it works.

**Comment by Michal-Leszczynski** [comment]
The `consistent_cluster_management` is set to true in used scylla.yaml. I got the same error even on a single dc setup.

It turned out that this error occurs only when running Scylla 5.4.0 with experimental `consistent-topology-changes`.

So should this issue be closed, or did `consistent-topology-changes` expose some flaws in alternator implementation that should be addressed?

**Comment by mykaul** [comment]
Looks like a basic bug - please provide all node logs and if there's a simple reproducer, that'd be great.

**Comment by mykaul** [comment]
(Note that in 5.4 this consistent topology is still experimental - until 6.0 or so)

**Comment by nyh** [comment]
> The `consistent_cluster_management` is set to true in used scylla.yaml. I got the same error even on a single dc setup.
> 
> It turned out that this error occurs only when running Scylla 5.4.0 with experimental `consistent-topology-changes`.

@kostja @kbr- it's confusing that we have two similarly named options, "consistent cluster management" and "consistent topology changes". How do they differ?

Running `test/alternator/run --raft` runs the entire Alternator test suite against Scylla running with `--consistent-cluster-management true`, and it works well. So the problem is NOT in `consistent-cluster-management`, as you already knew.

On master, adding `--experimental consistent-topology-changes` also didn't break any Alternator test. I also tried the same on 5.4.0, and again the Alternator tests all passed. Note that this tests all use a single node - it's possible that this `consistent-topology-changes` bug is specific to multi-node clusters, but @Michal-Leszczynski I don't know how to reproduce this problem so any chance that you can try to debug this or provide more hints when this problem happens?

@kbr- did we ever run dtests with the consistent-**topology**-changes flag? I'm not asking about running every day (I assume we don't), but just if we ever ran them.

**Comment by kostja** [comment]
* nyh ***@***.***> [23/12/11 12:55]:
> > The `consistent_cluster_management` is set to true in used scylla.yaml. I got the same error even on a single dc setup.
> > 
> > It turned out that this error occurs only when running Scylla 5.4.0 with experimental `consistent-topology-changes`.
> 
> @kostja @kbr- it's confusing that we have two similarly named options, "consistent cluster management" and "consistent topology changes". How do they differ?
> 
> Running `test/alternator/run --raft` runs the entire Alternator test suite against Scylla running with `--consistent-cluster-management true`, and it works well. So the problem is NOT in `consistent-cluster-management`, as you already knew.
> 

Yeah, I know, to my excuse:
- both names are checked with Tzach
- both names will fall into oblivion after some time (will become
  the non-changeable default).

> On master, adding `--experimental consistent-topology-changes` also didn't break any Alternator test. I also tried the same on 5.4.0, and again the Alternator tests all passed. Note that this tests all use a single node - it's possible that this `consistent-topology-changes` bug is specific to multi-node clusters, but @Michal-Leszczynski I don't know how to reproduce this problem so any chance that you can try to debug this or provide more hints when this problem happens?
> 
> @kbr- did we ever run dtests with the consistent-**topology**-changes flag? I'm not asking about running every day (I assume we don't), but just if we ever ran them.

We're currently in the process of building a map of failures,
please ping me on irc I'll share a link (overall reachable from
the topology plan in google docs).

-- 
Konstantin Osipov, Moscow, Russia

**Comment by Michal-Leszczynski** [comment]
> On master, adding --experimental consistent-topology-changes also didn't break any Alternator test. I also tried the same on 5.4.0, and again the Alternator tests all passed. Note that this tests all use a single node - it's possible that this consistent-topology-changes bug is specific to multi-node clusters, but @Michal-Leszczynski I don't know how to reproduce this problem so any chance that you can try to debug this or provide more hints when this problem happens?

Actually, I was mistaken about the `consistent-topology-changes` being the only cause. The root cause seems to be `consistent-topology-changes` enabled alongside `tablets` in experimental features. Sorry for misinformation. With those two experimental flags enabled, behavior described above happens even on a single node cluster by running script from alternator docs:
```python
import boto3
dynamodb = boto3.resource('dynamodb',endpoint_url='http://192.168.200.31:8000',
                  region_name='None', aws_access_key_id='None', aws_secret_access_key='None')

dynamodb.create_table(
    AttributeDefinitions=[
    {
        'AttributeName': 'key',
        'AttributeType': 'S'
    },
    ],
    BillingMode='PAY_PER_REQUEST',
    TableName='usertable',
    KeySchema=[
    {
        'AttributeName': 'key',
        'KeyType': 'HASH'
    },
    ])
```

node logs: [alternator.log](https://github.com/scylladb/scylladb/files/13648379/alternator.log)
scylla.yaml: [scyllayaml.zip](https://github.com/scylladb/scylladb/files/13648496/scyllayaml.zip)

**Comment by nyh** [comment]
I see. Please see https://github.com/scylladb/scylladb/issues/16203, already fixed by https://github.com/scylladb/scylladb/issues/16203 (not yet committed) - Alternator didn't really work correctly in some cases with tablets.

However, I never saw the specific problem that you saw, perhaps because I never enabled the experimental "consistent-topology-changes" (that thanks it its confusing name, I still have no idea what it does!). Can you please confirm that you don't get the same problem if you only enable **tablets** but not **consistent-topology-changes**?

**Comment by Michal-Leszczynski** [comment]
I don't think that it's possible to enable `tablets` without `consistent-topology-changes`:
```
ERROR 2023-12-12 15:30:12,044 [shard 0:main] init - Bad configuration: The consistent-topology-changes feature has to be enabled if tablets feature is enabled
```

**Comment by nyh** [comment]
Note that after https://github.com/scylladb/scylladb/issues/16203, it's possible to create an Alternator table with tablets turned on (but it won't use tablets by default) - and I have tests that check it. Before this patch goes in, there isn't much point of even trying.

So I think you should, please, check this again after that PR goes in, and if it works, I think this issue should be closed - I don't think there's any intention that actual users should use Scylla 5.4.0 with tablets and expect it to work correctly, right?

**Comment by kostja** [comment]
Won't fix in 5.4, addressed in trunk (or will be).

**Comment by nyh** [comment]
> Won't fix in 5.4, addressed in trunk (or will be).

@Michal-Leszczynski can you please try the same thing in Scylla master? I can't reproduce this problem, but I'm worried maybe what you tried is slightly different from what I tried (there was also some confusion above about what exactly you tried).

In general, if @kostja says that we won't be backporting any tablets-related fixes to 5.4, then I think it is pointless for you to be testing **anything** related to tablets on 5.4. If you want to experiment with tablets and prepare for it, you should be running an updated **nightly** version of Scylla. Tablets support is improving daily: For example, if you're trying Alternator, just last week GSIs didn't work on Alternator, and today they work.
